unixfs/src/dir/builder/buffered.rs

```diff
@@ -0,0 +1,390 @@
+    counter: u64,
+    opts: TreeOptions,
+}
+
+impl Default for BufferingTreeBuilder {
+    fn default() -> Self {
+        Self::new(TreeOptions::default())
+    }
+}
+
+impl BufferingTreeBuilder {
+    /// Construct a new tree builder with the given configuration.
+    pub fn new(opts: TreeOptions) -> Self {
+        BufferingTreeBuilder {
+            root_builder: DirBuilder::root(0),
+            longest_path: 0,
+            counter: 1,
+            opts,
+        }
+    }
+
+    /// Records the give path to be a link to the following cid.
+    ///
+    /// FIXME: this should be renamed as "put_leaf" or "put_opaque_leaf".
+    pub fn put_file(
+        &mut self,
+        full_path: &str,
+        target: Cid,
+        total_size: u64,
+    ) -> Result<(), TreeBuildingFailed> {
+        // inserted at the depth
+        let leaf = Leaf {
+            link: target,
+            total_size,
+        };
+
+        self.modify_with(full_path, |parent, basename, _| {
+            parent
+                .put_leaf(basename, leaf)
+                .map_err(|_| TreeBuildingFailed::DuplicatePath(full_path.to_string()))
+        })
+    }
+
+    /// Directories get "put" implicitly through the put files, and directories need to be adjusted
+    /// only when wanting them to have metadata.
+    pub fn set_metadata(
+        &mut self,
+        full_path: &str,
+        metadata: Metadata,
+    ) -> Result<(), TreeBuildingFailed> {
+        // create all paths along the way
+        //
+        // set if not set, error otherwise? FIXME: doesn't error atm
+        self.modify_with(full_path, |parent, basename, id| {
+            parent
+                .add_or_get_node(basename, id)
+                .map_err(|_| TreeBuildingFailed::LeafAsDirectory(full_path.to_string()))?
+                .set_metadata(metadata);
+            Ok(())
+        })
+    }
+
+    fn modify_with<F>(&mut self, full_path: &str, f: F) -> Result<(), TreeBuildingFailed>
+    where
+        F: FnOnce(&mut DirBuilder, String, &mut Option<u64>) -> Result<(), TreeBuildingFailed>,
+    {
+        // create all paths along the way
+        //
+        // assuming it's ok to split '/' since that cannot be escaped in linux at least
+
+        self.longest_path = full_path.len().max(self.longest_path);
+        let mut remaining = full_path.split('/').enumerate().peekable();
+        let mut dir_builder = &mut self.root_builder;
+
+        // needed to avoid borrowing into the DirBuilder::new calling closure
+        let counter = &mut self.counter;
+
+        while let Some((depth, next)) = remaining.next() {
+            let last = remaining.peek().is_none();
+
+            match (depth, next, last) {
+                // this might need to be accepted in case there is just a single file
+                (0, "", true) => { /* accepted */ }
+                (0, "", false) => {
+                    return Err(TreeBuildingFailed::RootedPath(full_path.to_string()))
```

Comment:
the test cases like `should add a directory from the file system` will of course use `/path` ... Perhaps it should just be stripped.

Reply:
Checked that jsipfs just strips the slash out if it exists and we could do the same here. Did not check if rooted and non-rooted paths can be mixed however.

Reply:
Kept the error, moved path stripping over to http level like in js-ipfs.